spring-web/src/main/java/org/springframework/http/client/reactive/JettyClientHttpRequest.java

```diff
@@ -0,0 +1,146 @@
+ *
+ * @author Sebastien Deleuze
+ * @since 5.1
+ */
+public class JettyClientHttpRequest extends AbstractClientHttpRequest {
+
+	private final Request jettyRequest;
+
+	private final DefaultDataBufferFactory bufferFactory = new DefaultDataBufferFactory();
+
+	private ReactiveRequest reactiveRequest;
+
+	private Mono<ClientHttpResponse> response;
+
+
+	public JettyClientHttpRequest(Request jettyRequest) {
+		this.jettyRequest = jettyRequest;
+	}
+
+
+	@Override
+	protected void applyHeaders() {
+		getHeaders().entrySet().forEach(e -> e.getValue().forEach(v -> this.jettyRequest.header(e.getKey(), v)));
+		if (!getHeaders().containsKey(HttpHeaders.ACCEPT)) {
+			this.jettyRequest.header(HttpHeaders.ACCEPT, "*/*");
+		}
+	}
+
+	@Override
+	protected void applyCookies() {
+		getCookies().values().stream().flatMap(Collection::stream)
+				.map(cookie -> new HttpCookie(cookie.getName(), cookie.getValue()))
+				.forEach(this.jettyRequest::cookie);
+	}
+
+	@Override
+	public HttpMethod getMethod() {
+		HttpMethod method = HttpMethod.resolve(this.jettyRequest.getMethod());
+		Assert.notNull(method, "Method must not be null");
+		return method;
+	}
+
+	@Override
+	public URI getURI() {
+		return this.jettyRequest.getURI();
+	}
+
+	@Override
+	public DataBufferFactory bufferFactory() {
+		return this.bufferFactory;
+	}
+
+	@Override
+	public Mono<Void> writeWith(Publisher<? extends DataBuffer> publisher) {
+		String contentType = jettyRequest.getHeaders().contains(HttpHeader.CONTENT_TYPE) ?
+				jettyRequest.getHeaders().getField(HttpHeader.CONTENT_TYPE).getValue():
+				MediaType.APPLICATION_OCTET_STREAM_VALUE;
+		Publisher<ContentChunk> chunks = Flux.from(publisher).map(DataBuffer::asByteBuffer).map(ContentChunk::new);
+		ReactiveRequest.Content requestContent = ReactiveRequest.Content.fromPublisher(chunks, contentType);
+		this.reactiveRequest = ReactiveRequest.newBuilder(jettyRequest).content(requestContent).build();
+		return doCommit(this::completes);
+	}
+
+	@Override
+	public Mono<Void> writeAndFlushWith(Publisher<? extends Publisher<? extends DataBuffer>> body) {
+		String contentType = this.jettyRequest.getHeaders().getField(HttpHeader.CONTENT_TYPE).getValue();
+		ReactiveRequest.Content content = ReactiveRequest.Content.fromPublisher(Flux
+				.from(body)
+				.flatMap(publisher -> publisher)
+				.map(buffer -> new ContentChunk(buffer.asByteBuffer())), contentType);
+		this.reactiveRequest = ReactiveRequest.newBuilder(this.jettyRequest).content(content).build();
+		return doCommit(this::completes);
+	}
+
+	@Override
+	public Mono<Void> setComplete() {
+		if (this.reactiveRequest == null) {
+			this.reactiveRequest = ReactiveRequest.newBuilder(this.jettyRequest).build();
+		}
+		return doCommit(this::completes);
+	}
+
+	private Mono<Void> completes() {
+		this.response = Mono.from(
+				this.reactiveRequest.response((reactiveResponse, contentChunks) -> {
+
+					// Implementation with buffer copy instead of optimized buffer wrapping
```

Comment:
Not exactly sure what is meant by this, let's discuss it during our meeting.

Reply:
With current Jetty client implementation, we can never transform a `Flux<DataBuffer>` to `Mono<DataBuffer>` because the callback calls `request(1)` + release the buffer. In our case we want to `request(1)` ( to get next buffer) during buffer composition but do the release only when `PooledDataBuffer#release` is called, which is impossible with current API. See the related https://github.com/eclipse/jetty.project/issues/2429 Jetty issue.